compiler-rt/lib/asan/asan_globals.cpp

```diff
@@ -147,14 +163,21 @@ static void CheckODRViolationViaIndicator(const Global *g) {
     *odr_indicator = REGISTERED;
     return;
   }
+
+  // Fetch globals with the same ODR indicator.
+  auto *relevant_globals_lookup =
+      map_of_globals_by_indicator.find(g->odr_indicator);
+  if (!relevant_globals_lookup)
+    return;
+
+  ListOfGlobals *relevant_globals = relevant_globals_lookup->second;
   // If *odr_indicator is DEFINED, some module have already registered
   // externally visible symbol with the same name. This is an ODR violation.
-  for (ListOfGlobals *l = list_of_all_globals; l; l = l->next) {
-    if (g->odr_indicator == l->g->odr_indicator &&
-        (flags()->detect_odr_violation >= 2 || g->size != l->g->size) &&
+  for (ListOfGlobals *l = relevant_globals; l; l = l->next) {
+    if ((flags()->detect_odr_violation >= 2 || g->size != l->g->size) &&
```

Comment:
Do we need full list?
why not to keep just the first global, and raise an error on the first mismatch?

Reply:
When `halt_on_error=1`, the first one's enough. If `halt_on_error=0`, my understanding is that keeping just the first one may make it report less violations. If keeping the existing behavior is not necessary, I can make it keep just the first global. We always halt on error, so for us just the first one would be enough. With that said, even when keeping the full list the perf was great compared to the current baseline.

LMK what you prefer.

Reply:
Taking into account we may still need unregistered